pymc3/distributions/distribution.py

```diff
@@ -134,40 +321,147 @@ def __new__(cls, name, *args, **kwargs):
+            # Assigning the testval earlier causes trouble because the RV may not be created with the final shape already.
+            rv_out.tag.test_value = testval
 
-        return model.register_rv(rv_out, name, data, total_size, dims=dims, transform=transform)
+        return model.register_rv(rv_out, name, observed, total_size, dims=dims, transform=transform)
 
     @classmethod
-    def dist(cls, dist_params, **kwargs):
+    def dist(
+        cls,
+        dist_params,
+        *,
+        shape: Optional[Shape] = None,
+        size: Optional[Size] = None,
+        testval=None,
+        **kwargs,
+    ) -> RandomVariable:
+        """Creates a RandomVariable corresponding to the `cls` distribution.
+
+        Parameters
+        ----------
+        dist_params : array-like
+            The inputs to the `RandomVariable` `Op`.
+        shape : int, tuple, Variable, optional
+            A tuple of sizes for each dimension of the new RV.
+
+            An Ellipsis (...) may be inserted in the last position to short-hand refer to
+            all the dimensions that the RV would get if no shape/size/dims were passed at all.
+        size : int, tuple, Variable, optional
+            For creating the RV like in Aesara/NumPy.
+        testval : optional
+            Test value to be attached to the output RV.
+            Must match its shape exactly.
 
-        testval = kwargs.pop("testval", None)
+        Returns
+        -------
+        rv : RandomVariable
+            The created RV.
+        """
+        if "dims" in kwargs:
+            raise NotImplementedError("The use of a `.dist(dims=...)` API is not supported.")
+        if shape is not None and size is not None:
+            raise ValueError(
+                f"Passing both `shape` ({shape}) and `size` ({size}) is not supported!"
+            )
+        shape = _convert_shape(shape)
+        size = _convert_size(size)
+
+        ndim_supp = cls.rv_op.ndim_supp
+        ndim_expected = None
+        ndim_batch = None
+        create_size = None
+
+        if shape is not None:
+            if Ellipsis in shape:
+                # Ellipsis short-hands all implied dimensions. Therefore
+                # we don't know how many dimensions to expect.
+                ndim_expected = ndim_batch = None
+                # Create the RV with its implied shape and resize later
+                create_size = None
+            else:
+                ndim_expected = len(tuple(shape))
+                ndim_batch = ndim_expected - ndim_supp
+                create_size = tuple(shape)[:ndim_batch]
+        elif size is not None:
+            ndim_expected = ndim_supp + len(tuple(size))
+            ndim_batch = ndim_expected - ndim_supp
+            create_size = size
+
+        # Create the RV with a `size` right away.
+        # This is not necessarily the final result.
+        rv_out = cls.rv_op(*dist_params, size=create_size, **kwargs)
+        ndim_actual = rv_out.ndim
```

Comment:
Should this be called `ndim_implied` as it is in the `__new__` method, or is it actually used differently?

Reply:
I'd rather rename the other one to `ndim_actual`, because both can include extra dimensions that are not implied by distribution support or parameters.

Reply:
Sounds good